### ISSUE TITLE
Access value outside lua filter by body block

### ISSUE THREAD

**Issue by vsingh63** [openresty/lua-nginx-module#1440]
I am using lua-nginx module v0.10.14 on NGINX server v1.12.2. I initialized one variable inside the location and then assigned it some value in body_filter_by_lua_block and trying to access it outside lua block. But, somehow it's value is not coming. is there something I'm missing here?

```
location / {
    proxy_pass someurl
    set xyz '';
    body_filter_by_lua_block {
        local vlaue, err = ngx.re.match("helloworld 1234", ([0-9]+), "io") 
        ngx.var.xyz = value[0] 
    }
echo $xyz //coming blank shouldn't be 1234??
}
```

**Comment by dhlin** [comment]
@vsingh63 you need to do this instead, check the doc under [echo-nginx-module](https://github.com/openresty/echo-nginx-module#filter-directives)
```
location / {
    proxy_pass someurl;
    set $xyz '';
    body_filter_by_lua_block {
        ngx.var.xyz = '1234'
    }
    echo_after_body $xyz;
}
```

**Comment by vsingh63** [comment]
@dhlin - My concern is not the echo the value though I'll try the after body directive. What I'm trying to do is that access the value of xyz outside the Lua block and do some calculation on it.

**Comment by tokers** [comment]
@vsingh63 
What you mixed is the directives execution priority. 
`echo` runs at content phase which is earlier than the filter phase (where `body_filter_by_lua_block` works).

> What I'm trying to do is that access the value of xyz outside the Lua block and do some calculation on it.

You may try to postpone the calculation into the log phase, combined with `log_by_lua*` directives cluster.

**Comment by vsingh63** [comment]
@tokers - As I said I'm not concerned on echoi I wrote that for explaining the things, to be specific I want to use this value to create one header in nginx. So codesnippet is something like this, but as I said the variable value is coming as empty hence header is not being set up. Is this also because of priority of directives.

```
location / {
    proxy_pass someurl
    set xyz '';
    body_filter_by_lua_block {
        local vlaue, err = ngx.re.match("helloworld 1234", ([0-9]+), "io") 
        ngx.var.xyz = value[0] 
    }
    add_header "custom-header" $xyz;
}
```

**Comment by tokers** [comment]
@vsingh63 
> As I said I'm not concerned on echoi I wrote that for explaining the things, to be specific I want to use this value to create one header in nginx. So codesnippet is something like this, but as I said the variable value is coming as empty hence header is not being set up. Is this also because of priority of directives.

Yes. Response headers already sent before the body generated. If possible, you can try to use the `ngx.location.capture`, spawn a subrequest in an earlier phase (like content phase), then analysis subrequest's body and set the header that you want.

**Comment by vsingh63** [comment]
@tokers - I meant nginx header using nginx **add_heade**r module which I hope executed after lua module. Also, I can't use subrequest. is there any other way to do it?
is variable initialized in body block is not accessible at all as ngx variable?

**Comment by vsingh63** [comment]
> @vsingh63
> 
> > As I said I'm not concerned on echoi I wrote that for explaining the things, to be specific I want to use this value to create one header in nginx. So codesnippet is something like this, but as I said the variable value is coming as empty hence header is not being set up. Is this also because of priority of directives.
> 
> Yes. Response headers already sent before the body generated. If possible, you can try to use the `ngx.location.capture`, spawn a subrequest in an earlier phase (like content phase), then analysis subrequest's body and set the header that you want.

@tokers @agentzh  - I tried it with the subsequent request and I'm able to get the response but that response body is not coming back to the browser/curl though response code coming is 200. Below is the code snippet

```
location /main/ {
lua_need_body_request_on;
content_by_lua_block {
       local res = ngx.location.capture("/sub/")
       ngx.print("resp::", res) // printing the response
}
}

location /sub/ {
    proxy_pass someurl
}
```

**Comment by vsingh63** [comment]
second approach working now, I forgot to add ngx.say(res.body).
Thanks!

**Comment by tokers** [comment]
@vsingh63 
`ngx.location.capture` captures the whole body into memory, so be careful when the body is huge.
